services/error.go

```diff
@@ -73,88 +78,76 @@ type StorageError struct {
 	}
 	return fmt.Sprintf("%s: %s, %s: %s", e.Op, e.Storager, e.Path, e.Err.Error())
 }
 
 // Unwrap implements xerrors.Wrapper
-func (e *StorageError) Unwrap() error {
+func (e StorageError) Unwrap() error {
 	return e.Err
 }
 
-// NewMetadataNotRecognizedError will create a new MetadataUnrecognizedError.
-func NewMetadataNotRecognizedError(key string, value interface{}) *MetadataUnrecognizedError {
-	return &MetadataUnrecognizedError{
-		Err:   ErrCapabilityInsufficient,
-		Key:   key,
-		Value: value,
-	}
-}
-
 // MetadataUnrecognizedError means this operation meets unrecognized metadata.
 type MetadataUnrecognizedError struct {
-	Err error
-
 	Key   string
 	Value interface{}
 }
 
-func (e *MetadataUnrecognizedError) Error() string {
-	return fmt.Sprintf("metadata unrecognized, %s, %v: %s", e.Key, e.Value, e.Err.Error())
+func (e MetadataUnrecognizedError) Error() string {
+	return fmt.Sprintf("metadata unrecognized, %s, %v: %s", e.Key, e.Value, ErrCapabilityInsufficient.Error())
 }
 
 // Unwrap implements xerrors.Wrapper
-func (e *MetadataUnrecognizedError) Unwrap() error {
-	return e.Err
-}
-
-// NewPairUnsupportedError will create a new PairUnsupportedError.
-func NewPairUnsupportedError(pair types.Pair) *PairUnsupportedError {
-	return &PairUnsupportedError{
-		Err:  ErrCapabilityInsufficient,
-		Pair: pair,
-	}
+func (e MetadataUnrecognizedError) Unwrap() error {
+	return ErrCapabilityInsufficient
 }
 
 // PairUnsupportedError means this operation has unsupported pair.
 type PairUnsupportedError struct {
-	Err error
-
 	Pair types.Pair
 }
 
-func (e *PairUnsupportedError) Error() string {
-	return fmt.Sprintf("pair unsupported, %s: %s", e.Pair, e.Err.Error())
+func (e PairUnsupportedError) Error() string {
+	return fmt.Sprintf("pair unsupported, %s: %s", e.Pair, ErrCapabilityInsufficient.Error())
 }
 
 // Unwrap implements xerrors.Wrapper
-func (e *PairUnsupportedError) Unwrap() error {
-	return e.Err
-}
-
-// NewPairRequiredError will create a new PairRequiredError.
-func NewPairRequiredError(keys ...string) *PairRequiredError {
-	return &PairRequiredError{
-		Err:  ErrRestrictionDissatisfied,
-		Keys: keys,
-	}
+func (e PairUnsupportedError) Unwrap() error {
+	return ErrCapabilityInsufficient
 }
 
 // PairRequiredError means this operation has required pair but missing.
 type PairRequiredError struct {
-	Err error
-
 	Keys []string
 }
 
-func (e *PairRequiredError) Error() string {
-	return fmt.Sprintf("pair required, %v: %s", e.Keys, e.Err.Error())
+func (e PairRequiredError) Error() string {
+	return fmt.Sprintf("pair required, %v: %s", e.Keys, ErrRestrictionDissatisfied.Error())
 }
 
 // Unwrap implements xerrors.Wrapper
-func (e *PairRequiredError) Unwrap() error {
-	return e.Err
+func (e PairRequiredError) Unwrap() error {
+	return ErrRestrictionDissatisfied
+}
+
+// ObjectModeInvalidError means the provided object mode is invalid.
+type ObjectModeInvalidError struct {
```

Comment:
It's a bit wired for an error struct that can't unwrap to an error code.

- Makes it hard for the user to discover which error go-storage could return.
- Makes it hard to check the error code: the user needs to use `errors.Is` & `errors.As` at the same time. What if user just wants to know what error happened and doesn't care about why and related context?

Maybe we should provide an error code for error struct (even when they are already the most underlying error)?

Reply:
I agree. If every error has an error code, it will be easier to understand.
Just like HTTP response codes can have 0, 1 or more error codes, our error codes can have 0, 1 or more error `struct`s?

Reply:
Yes, our error codes could be used as underlying errors for different error structs.

- error code represents `What` -> only a label to explain why the error happened
- error struct represents `Why` -> carrying context info to explain why the error happened